README.md
# COS301-DND-Backend

| Branch  |                                                                               Status                                                                               |
| ------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| Master  | [![Build Status](https://travis-ci.org/COS301-OptimizePrime/COS301_DND_Backend.svg?branch=master)](https://travis-ci.org/COS301-OptimizePrime/COS301_DND_Backend)  ![deployment diagram](https://img.shields.io/badge/coverage-82%25-brightgreen.svg)|
| Develop | [![Build Status](https://travis-ci.org/COS301-OptimizePrime/COS301_DND_Backend.svg?branch=develop)](https://travis-ci.org/COS301-OptimizePrime/COS301_DND_Backend) ![deployment diagram](https://img.shields.io/badge/coverage-82%25-brightgreen.svg) |

Backed API Server

## Deployment Diagram
![deployment diagram](https://i.imgur.com/DFwdbp0.jpg)

## Features

- HTTP/2 using gRPC (https://grpc.io/)
- ~~HTTPS in production mode~~
- Real time sockets
- Firebase authentication
- Strongly typed service and message definition (Protobuf3)
- Easy configuration with TOML
- pep8 Compliance
- Regression testing
- Developer mode (SQLite)
- Production mode (PostgreSQL and PyPy3)
- SystemD service file


## Coverage

|Name                          |Stmts  | Miss| Cover|
|:-----------------------------|-----:|----:|-----:|
|main.py                 |  50    |  5   | 90%|
|server/character.py     | 281    | 73   | 74%|
|server/db.py            | 167    |  3   | 98%|
|server/firebase.py      |   4    |  0   |100%|
|server/helpers.py       | 181    |  1   | 99%|
|server/log.py           |  18    |  0   |100%|
|server/server_pb2.py    | 186    |  0   |100%|
|server/session.py       | 643    |200   | 69%|
|TOTAL                   |1530    |282   | 82%|


## Requirements

```bash
sudo apt-get update
sudo apt-get install python3
sudo apt-get install pip3
sudo apt-get install nodejs
sudo apt-get install dart
```

## Installing packages

```bash
make install
```

## Running server

```bash
make run
```

## Running the tests

```bash
make test
```

## Generating code from protobuffers

```bash
python -m grpc.tools.protoc -I./protos --python_out=./server --grpc_python_out=./server ./protos/server.proto
```

## Fedora Installation
```bash
sudo dnf update
sudo dnf install postgresql-server postgresql-contrib
sudo systemctl enable postgresql
sudo postgresql-setup --initdb --unit postgresql
sudo systemctl start postgresql
pypy -m ensurepip
```

## Arch Linux Installation
```bash
pacman -S sudo
sudo pacaman -Syyu
sudo pacman -S postgresql

sudo -u postgres -i
initdb --locale en_US.UTF-8 -E UTF8 -D '/var/lib/postgres/data'
exit

sudo systemctl enable postgresql
sudo systemctl start postgresql
sudo -u postgres -i

createuser --interactive
```

ChangeLog
CHANGES
=======

* Added pbr to requirements. Fixed tests
* Fixed sessions -> light\_sessions
* Added new meta files
* Bug fixes. Allow more sessions in development mode. Added proper package meta data
* Massive changes to the API. Implemented Character and session relationships
* Restructure project. Database should now not close after each call to opitimise usage. (Need to add long runnning tests). Started to experiment with PyPy. Started to add version info
* Updated all exceptions to print sql error to log. Updted listSessions to have state included
* Removed lock file
* Added gender and level attributes to character. Closes #41
* Ready functionality should be working and state change functionality
* Started adding ready up functionality
* autopep8 formatting
* Fixed bug with character replies. (status message\_
* added features and traits
* Added session id to character. Fixed bug in updating character. Added test to check if character name is blank
* Added back session tests
* Added equipment for characters. Added tests for equipment. Fixed bug where old relationship rows are not deleted (cascading)
* Updated dart library. Added pip upgrade to makefile for python modules
* Added error handling on edge cases for database
* Added more fields for character fields
* Added hitpoints and savingthrows tests
* Added more tests. Various bug fixes
* Added more helper rpc functions for characters. Incomplete conversion between objects will finish tommorow
* Multiple bug fixes for characters. More tests
* Close connections to database
* close database connections
* formatting. desc sorting
* Added ordering to relationships. DM sessions should now be listed
* first semi working characters service. Need to write tests. Need to check performance
* removed update npm and pip
* updates to travis
* added updates to makefile
* changes to travis
* Started adding character prototypes
* fixed syntax error
* more changes for python 2
* Changes for python 2
* Added configuration
* Formatted all files with pep8. Added a extra benchmark test. Fixed assertions
* Fixed get session bug. List sessions are also not updating the user's name. (Fixed)
* forgot to add the actual test file
* Modified test script to see if it works on travis
* Changed testing. Added ability to get all the sessions of a user. Reuse connections when possible
* Fixed a bug where joining the same session multiple times will make the session appear full
* added gracefull stopping
* Fixed service
* Added service files. Fixed default size of a new session

v0.0.1
------

* Added ability to switch to Postgresql for production enviroments. Do have some race conditions. Closes #7
* Implemented limits for amount of sessions users can have. Closes #16
* Implemented SetPrivate RPC. Closes #15
* Removed duplicate code
* Implemented SetName RPC. Closes #14
* Leaving a session with no other players deletes the session. Fixes #10. If the Dungeon master leaves a new dungeon master is assigned
* Fixes #13
* Added privacy settings for session
* Fixed #12, Completed #11
* Added the ability to list sessions that are full based on parameter
* List sessions now return the most recently created sessions (desc)
* removed bin as it is replaced by actual tests for example code
* Added test that users can not modify sessions if they are not the dungeon master
* added get session by id. tests for if users try to leave a session they are not in
* need to add tests to see if session is empty. also need to do getSessionByID
* added leave session rpc
* added tests for join
* removed cache files
* Updated readme
* added module file
* added SQLAlchemy to requirements
* Moved to SQLAlchemy need to write additional tests. #1
* travis
* moved all installation to makefile
* Added dart to travis
* Started adding dart tests for the API
* New protocol definitions
* Started adding functionality for set max player
* add build status to readme
* hopefully this works on travis
* fixed makefile for travis
* removed travis python specifiers
* switch back to default python to help with travis loading times
* fixes
* attempt to use pip3
* updated to reflect python3
* removed lock file
* added node requirements
* changes so travis doesnt get stuck
* added encrypted file for travis
* travis updates
* travis updates
* removed bad pip install
* travis fixes
* started adding travis
* Added list sessions rpc
* removed outdated dart libraries
* protocol updates
* Added proper login tests using nodejs. Will need to build a proper install setup for this
* removed
* removed
* Added some basic tests. Hard to actually do with auth included
* Added links to the readme
* Working demo for how a client will be implemented
* Removed lock file
* Generated complementary dart library
* Trying to generate dart code for client
* Initial protocol buffers and gRPC implementation
* Initial commit


server/db.py
import datetime
import os

from sqlalchemy import (Boolean, Column, DateTime, ForeignKey, Integer,
                        SmallInteger, String, Table, create_engine)
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, scoped_session, sessionmaker

from . import config

Base = declarative_base()


def connect():
    if os.environ['ENV'] == 'prod':
        # logger = logging.getLogger('cos301-DND')
        # logger.debug('Using PostgreSQL!')
        engine = create_engine('postgresql+psycopg2cffi://' +
                               str(config.val['database']['username']) +
                               ':' +
                               str(config.val['database']['password']) +
                               '@' +
                               str(config.val['database']['address']) +
                               ':' +
                               str(config.val['database']['port']) +
                               '/dnd_backend')
    else:
        engine = create_engine(
            'sqlite:///./dnd_backend.db',
            echo=False,
            connect_args={
                'check_same_thread': False})

    Base.metadata.create_all(engine)

    session_factory = sessionmaker(bind=engine)
    session = scoped_session(session_factory)

    return session

user_sessions = Table('usersessions', Base.metadata,
                      Column('users_id', Integer, ForeignKey('users.id')),
                      Column('sessions_id', Integer, ForeignKey('sessions.id'))
                      )

user_ready_sessions = Table('userreadysession', Base.metadata,
                            Column('users_id', Integer, ForeignKey('users.id')),
                            Column('sessions_id', Integer, ForeignKey('sessions.id'))
                            )

class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    uid = Column(String, nullable=False, index=True)
    name = Column(String, nullable=False)
    date_created = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now)

    date_updated = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now,
        onupdate=datetime.datetime.now)

    # User can have many sessions where the user is a dungeon master
    session_dungeon_masters = relationship(
        "Session",
        back_populates="dungeon_master",
        order_by="desc(Session.date_created)")

    # User has many sessions
    joined_sessions = relationship(
        "Session",
        secondary=user_sessions,
        order_by="desc(Session.date_created)")

    # User has many characters
    characters = relationship(
        "Character", back_populates="creator")

    # User can be ready in multiple sessions although this is prevented by the program logic.
    ready_in_session = relationship(
        "Session",
        secondary=user_ready_sessions,
        order_by="desc(Session.date_created)")

    # Last socket
    # socket = Column(String, nullable=False)

    def __repr__(self):
        return "<User(id='%s', uid='%s', name='%s')>" % (
            self.id, self.uid, self.name)


class Session(Base):
    __tablename__ = 'sessions'

    id = Column(Integer, primary_key=True)
    session_id = Column(String(36), nullable=False, index=True)
    name = Column(String(100), nullable=False)
    date_created = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now)

    date_updated = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now,
        onupdate=datetime.datetime.now)

    max_players = Column(SmallInteger, nullable=False)
    full = Column(Boolean, nullable=False, default=False)
    private = Column(Boolean, nullable=False, default=False)

    ready_up_expiry_time = Column(SmallInteger, nullable=False)

    state = Column(String(50), nullable=False)
    state_ready_start_time = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now)
    state_meta = Column(Integer, nullable=False)

    # Session can have many ready users.
    ready_users = relationship(
        "User",
        secondary=user_ready_sessions,
        back_populates="ready_in_session")

    # Session has only one dungeon master
    dungeon_master_id = Column(Integer, ForeignKey('users.id'))
    dungeon_master = relationship(
        "User", back_populates="session_dungeon_masters")

    # Session has many users
    users_in_session = relationship(
        "User",
        secondary=user_sessions,
        back_populates="joined_sessions")

    users = association_proxy('users_in_session', 'user')

    characters_in_session = relationship(
        "Character",
        back_populates="session"
    )

    def __repr__(self):
        return "<Session(id='%s', session_id='%s', name='%s', dungeon_master='%s')>" % (
            self.id, self.session_id, self.name, self.dungeon_master)


class Character(Base):
    __tablename__ = 'characters'

    id = Column(Integer, primary_key=True)
    character_id = Column(String(36), nullable=False, index=True)
    name = Column(String(100), nullable=False)
    date_created = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now)

    date_updated = Column(
        DateTime,
        nullable=False,
        default=datetime.datetime.now,
        onupdate=datetime.datetime.now)

    # Character has only one creator/user
    creator_id = Column(Integer, ForeignKey('users.id'))
    creator = relationship(
        "User", back_populates="characters")

    session_id = Column(Integer, ForeignKey('sessions.id'))
    session = relationship(
        "Session",
        back_populates="characters_in_session"
    )

    skills = relationship(
        "Skill",
        uselist=False,
        back_populates="character",
        cascade="all, delete-orphan")

    attacks_and_spellcasting = relationship(
        "Attacks_And_Spellcasting",
        uselist=False,
        back_populates="character",
        cascade="all, delete-orphan")

    hitpoints = relationship(
        "Hitpoints",
        uselist=False,
        back_populates="character",
        cascade="all, delete-orphan")

    equipment = relationship(
        "Equipment",
        back_populates="character",
        cascade="all, delete-orphan")

    strength = Column(Integer, nullable=False)
    strength_subscript = Column(Integer, nullable=False, default=False)

    dexterity = Column(Integer, nullable=False)
    dexterity_subscript = Column(Integer, nullable=False, default=False)

    constitution = Column(Integer, nullable=False)
    constitution_subscript = Column(Integer, nullable=False, default=False)

    intelligence = Column(Integer, nullable=False)
    intelligence_subscript = Column(Integer, nullable=False, default=False)

    wisdom = Column(Integer, nullable=False)
    wisdom_subscript = Column(Integer, nullable=False, default=False)

    charisma = Column(Integer, nullable=False)
    charisma_subscript = Column(Integer, nullable=False, default=False)

    character_class = Column(String(100), nullable=False)
    race = Column(String(100), nullable=False)

    xp = Column(Integer, nullable=False)
    alignment = Column(String(200), nullable=False)
    background = Column(String(200), nullable=False)

    inspiration = Column(Integer, nullable=False)
    proficiency_bonus = Column(Integer, nullable=False)

    passive_wisdom = Column(Integer, nullable=False)

    personality_traits = Column(String(200), nullable=False)
    ideals = Column(String(200), nullable=False)
    bonds = Column(String(200), nullable=False)
    flaws = Column(String(200), nullable=False)

    features_and_traits = Column(String(350), nullable=False)

    gender = Column(String(50), nullable=False)
    level = Column(Integer, nullable=False)

class Skill(Base):
    __tablename__ = 'skills'

    id = Column(Integer, primary_key=True)
    character_id = Column(Integer, ForeignKey('characters.id'))
    character = relationship("Character", back_populates="skills")

    acrobatics = Column(Integer, nullable=False)
    acrobatics_proficient = Column(Boolean, nullable=False, default=False)

    animal_handling = Column(Integer, nullable=False)
    animal_handling_proficient = Column(Boolean, nullable=False, default=False)

    arcana = Column(Integer, nullable=False)
    arcana_proficient = Column(Boolean, nullable=False, default=False)

    athletics = Column(Integer, nullable=False)
    athletics_proficient = Column(Boolean, nullable=False, default=False)

    deception = Column(Integer, nullable=False)
    deception_proficient = Column(Boolean, nullable=False, default=False)

    history = Column(Integer, nullable=False)
    history_proficient = Column(Boolean, nullable=False, default=False)

    insight = Column(Integer, nullable=False)
    insight_proficient = Column(Boolean, nullable=False, default=False)

    intimidation = Column(Integer, nullable=False)
    intimidation_proficient = Column(Boolean, nullable=False, default=False)

    investigation = Column(Integer, nullable=False)
    investigation_proficient = Column(Boolean, nullable=False, default=False)

    medicine = Column(Integer, nullable=False)
    medicine_proficient = Column(Boolean, nullable=False, default=False)

    nature = Column(Integer, nullable=False)
    nature_proficient = Column(Boolean, nullable=False, default=False)

    perception = Column(Integer, nullable=False)
    perception_proficient = Column(Boolean, nullable=False, default=False)

    performance = Column(Integer, nullable=False)
    performance_proficient = Column(Boolean, nullable=False, default=False)

    persuasion = Column(Integer, nullable=False)
    persuasion_proficient = Column(Boolean, nullable=False, default=False)

    religion = Column(Integer, nullable=False)
    religion_proficient = Column(Boolean, nullable=False, default=False)

    sleight_of_hand = Column(Integer, nullable=False)
    sleight_of_hand_proficient = Column(Boolean, nullable=False, default=False)

    stealth = Column(Integer, nullable=False)
    stealth_proficient = Column(Boolean, nullable=False, default=False)

    survival = Column(Integer, nullable=False)
    survival_proficient = Column(Boolean, nullable=False, default=False)


class Attacks_And_Spellcasting(Base):
    __tablename__ = 'attacks_and_spellcasting'

    id = Column(Integer, primary_key=True)
    character_id = Column(Integer, ForeignKey('characters.id'))
    character = relationship("Character",
                             back_populates="attacks_and_spellcasting")

    name_1 = Column(String(100), nullable=False)
    name_2 = Column(String(100), nullable=False)
    name_3 = Column(String(100), nullable=False)
    atk_bonus_1 = Column(Integer, nullable=False)
    atk_bonus_2 = Column(Integer, nullable=False)
    atk_bonus_3 = Column(Integer, nullable=False)
    dmage_type_1 = Column(String(100), nullable=False)
    dmage_type_2 = Column(String(100), nullable=False)
    dmage_type_3 = Column(String(100), nullable=False)


class Hitpoints(Base):
    __tablename__ = 'hitpoints'

    id = Column(Integer, primary_key=True)
    character_id = Column(Integer, ForeignKey('characters.id'))
    character = relationship("Character", back_populates="hitpoints")

    armor_class = Column(Integer, nullable=False)
    current_hitpoints = Column(Integer, nullable=False)
    max_hitpoints = Column(Integer, nullable=False)
    temporary_hitpoints = Column(Integer, nullable=False)
    hitdice = Column(String(100), nullable=False)


class Equipment(Base):
    __tablename__ = 'equipment'

    id = Column(Integer, primary_key=True)
    character_id = Column(Integer, ForeignKey('characters.id'))
    character = relationship("Character", back_populates="equipment")

    name = Column(String(100), nullable=False)
    value = Column(Integer, nullable=False)